### PR TITLE
Introduce geometry primitives

### PR DESCRIPTION
**Pull request by Caellian**
# Changes

This PR introduces `point` and `rect` primitives and replaces previous value pairs with them. It also replaces output-wayland.cc `rectangle` as it's effectively doing the same.

This aims to make future changes more consistent and to reduce number of global variables by grouping them. As a side bonus, it allows us to optimize some common operations a little by relying on [Vc](https://vcdevel.github.io/Vc-1.4.4/) to make the compiler more aware of available opportunities to parallelize computation. Just grouping things results in significant reduction in number of generated move instructions:

image

## Details

- [x] Added `vec` and `rect` structures.
- [x] Added [Vc](https://vcdevel.github.io/Vc-1.4.4/) dependency for better SIMD optimization of common vector  operations.
  - This was previously done through Eigen, but I removed Eigen as 90% of its functionality wouldn't ever be used by Conky so it only added to maintenance burdain.

### CODE REVIEW

**Comment by Caellian** [comment]
@brndnmtthws Is it okay to add Eigen as dependency? I can do without it, but it might optimize some instructions. Alternatively, I could make it optional and apply SIMD only if present.

File: src/display-x11.cc [hunk 1]
```diff
@@ -648,26 +646,28 @@ bool handle_event<x_event_handler::CONFIGURE>(
...
+      text_width = new_text_size.x();
+      text_height = new_text_size.y();
 
-      text_width = window.width - 2 * border_total;
-      text_height = window.height - 2 * border_total;
-      int mw = surface->dpi_scale(maximum_width.get(*state));
-      if (text_width > mw && mw > 0) { text_width = mw; }
+      // don't apply dpi scaling to max pixel size
+      int mw = std::max(maximum_width.get(*state), 0);
+      if (text_width > mw) { text_width = mw; }
```
Comment (Caellian) [inline src/display-x11.cc:661]: Fixes #1528 hopefully

File: src/x11.cc [hunk 2]
```diff
@@ -780,14 +781,14 @@ void x11_init_window(lua::state &l, bool own) {
...
+
   int64_t input_mask = ExposureMask | PropertyChangeMask;
 #ifdef OWN_WINDOW
-  if (own_window.get(l)) {
-    input_mask |= StructureNotifyMask;
 #if !defined(BUILD_XINPUT)
-    input_mask |= ButtonPressMask | ButtonReleaseMask;
+  if (own_window.get(l)) { input_mask |= ButtonPressMask | ButtonReleaseMask; }
 #endif
-  }
```
Comment (Caellian) [inline src/x11.cc:790]: Did this too soon. Anyway, this is supposed to add window change watching in one of future commits to update window bounds caching and fix #1852

**Review by Caellian — COMMENTED** [review]
Last commit is expected be functionally equivalent to original

**Comment by brndnmtthws** [comment]
> @brndnmtthws Is it okay to add Eigen as dependency? I can do without it, but it might optimize some instructions. Alternatively, I could make it optional and apply SIMD only if present.
> 
> EDIT: We do surprisingly many vector computations while drawing components. I don't use majority of Eigen though. It might be better to simply add a way to provide optimized assembly for certain architectures and data types. 🤷‍♂️ And then someone can easily add optimizations later if they want to.

I haven't used Eigen, but I suppose if we can include it as a 3rd party dependency in the source tree (like we do with tolua++ and catch2), that would be fine. Hopefully it doesn't increase compile times too much. I would put it alongside toluapp in the `3rdparty/` dir.

**Comment by brndnmtthws** [comment]
I pushed a branch here to demonstrate what I meant by including it in the `3rdparty/` dir: https://github.com/brndnmtthws/conky/tree/add-eigen-3rdparty

You should be able to just cherry-pick e84f7b65635c7419742206b2928205759928605b from that branch, but remove the demo code from `main.cc`.

**Comment by Caellian** [comment]
> I pushed a branch here to demonstrate what I meant by including it in the `3rdparty/` dir: https://github.com/brndnmtthws/conky/tree/add-eigen-3rdparty
> 
> You should be able to just cherry-pick [e84f7b6](https://github.com/brndnmtthws/conky/commit/e84f7b65635c7419742206b2928205759928605b) from that branch, but remove the demo code from `main.cc`.

I added it as a submodule and made CMake update submodules during configure. IMO this is better because:
- we're not tracking entire eigen sources in our history
- it's much easier to update it
- support for BLAS and LaPACK can be configured while building conky
- pulling all the sources would make this PR huge

 I got it to build correctly, just have to figure out how to include the target as dependency.

**Comment by brndnmtthws** [comment]
> I added it as a submodule and made CMake update submodules during configure. IMO this is better because:
> 
>     * we're not tracking entire eigen sources in our history
> 
>     * it's much easier to update it
> 
>     * support for BLAS and LaPACK can be configured while building conky
> 
>     * pulling all the sources would make this PR huge
> 
>     * source authenticity doesn't have to be validated if someone's feeling paranoid 😄
> 
> 
> I got it to build correctly (takes ~1s to configure; build time is the same), just have to figure out how to include the target as dependency.
> 
> I still got little work to do on this PR anyway, so it will probably be done in a day or so.

I would prefer not to introduce a submodule. Submodules are usually more trouble than their worth. It would be much easier to just pick a specific release and include the sources we need. There's also subtrees, but I'm guessing those aren't worth the effort either.

**Comment by Caellian** [comment]
> Submodules are usually more trouble than they're worth.

I think they're the most ergonomic solution in this case because we're not going to be modifying eigen sources, and build works even with shallow clone due to the [script](https://gist.github.com/Caellian/9f92506fa24df111729589d392e84553) I added. So no one would need to interact with the submodule unless it's a version bump, which would've been much simpler than reapplying modifications on updates. I personally see no discrete reason to not use a submodule in this case.

**Comment by brndnmtthws** [comment]
> I think they're the most ergonomic solution in this case because we're not going to be modifying eigen sources, and build works even with shallow clone due to the [script](https://gist.github.com/Caellian/9f92506fa24df111729589d392e84553) I added. So no one would need to interact with the submodule unless it's a version bump, which would've been much simpler than reapplying modifications on updates. I personally see no discrete reason to not use a submodule in this case.

I understand your perspective, but it just isn't necessary to introduce all that complexity for something that doesn't require it. If we want to upgrade Eigen in the future, it's as simple as updating the source files, but for the most part we can treat 3rd party code as immutable, and layering everything on top of submodules just introduces unneeded complexity. I don't want downstream consumers or package maintainers to have to deal with all the quirks and edge cases that will inevitably be introduced.

**Review by brndnmtthws — APPROVED** [review]
👍 , seems like Vc is a better choice than Eigen for our needs.

This is a pretty big change, so it's probably going to have some unintended side effects. There are a few things I'd refactor, but probably better to do it in a separate PR later.

**Comment by Caellian** [comment]
> There are a few things I'd refactor, but probably better to do it in a separate PR later.

Same here :laughing:, I just tried adding semantics to rectangle via tempates (absolute vs. sized) and realized there's a lot that could be improved. C++ doesn't have traits as Rust does so it's somewhat tricky to get things like swizzling to layer properly and still allow the compiler to optimize code.

I'll probably address those issues at some point, but it's better to have _something_ and I don't want to keep this PR waiting for too long as it will be painful to update.

**Pull request merged**